File: BFS/Basics/127. Word Ladder/127. Word Ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        
        Set<String> q1 = new HashSet<>();
        Set<String> q2 = new HashSet<>();
        Set<String> visited = new HashSet<>();
        Set<String> hashSet = new HashSet<>(wordList);
        
        if (!wordList.contains(endWord)) {
            return 0;
        }
        
        q1.add(beginWord);
        q2.add(endWord);
        
        int step = 0;
        while (!q1.isEmpty() && !q2.isEmpty()) {
            
            step++;
            if (q1.size() > q2.size()) {
                Set<String> temp = q1;
                q1 = q2;
                q2 = temp;
            }
            
            Set<String> q = new HashSet<>();
            for (String curWord : q1) {
                
                if (visited.contains(curWord)) continue;
                visited.add(curWord);
                
                if (q2.contains(curWord)) {
                    return step;
                }
                
                for (int i = 0; i < curWord.length(); i++) {
                    
                    char[] word = curWord.toCharArray();
                    for (char j = 'a'; j <= 'z'; j++) {
                        if (curWord.charAt(i) == j) continue;
                        word[i] = j;
                        String tempWord = String.valueOf(word);
                        if (!hashSet.contains(tempWord)) continue;
                        if (visited.contains(tempWord)) continue;
                        q.add(tempWord);
                    }
                }
            }  
            q1 = q;
        }
        return 0;
    }
}

File: BFS/Basics/127. Word Ladder/127. Word Ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        
        unordered_map<string, int> visited;
        unordered_set<string> Set(wordList.begin(), wordList.end());
        if (Set.find(endWord) == Set.end()) {
            return 0;
        }
        
        unordered_set<string> q1;
        unordered_set<string> q2;
        q1.insert(beginWord);
        q2.insert(endWord);
        
        int step = 0;
        while (!q1.empty() && !q2.empty()) {
            
            step++;
            if (q1.size() > q2.size()) {
                std::swap(q1, q2);
            }
            
            unordered_set<string> q;
            for (auto curWord : q1) {
                
                if (visited[curWord]) continue;
                visited[curWord] = 1;
                
                if (q2.find(curWord) != q2.end()) {
                    return step;
                }
                
                for (int i = 0; i < curWord.size(); i++) {
                    
                    char temp = curWord[i];
                    for (char j = 'a'; j <= 'z'; j++) {
                        curWord[i] = j;
                        if (Set.find(curWord) == Set.end()) continue;
                        if (visited[curWord]) continue;
                        q.insert(curWord);
                    }
                    curWord[i] = temp;
                }
            }
            std::swap(q, q1);
        }
        return 0;
    }
};